### ISSUE TITLE
Consider implicit casting of single pointers to many pointers

### ISSUE THREAD

**Issue by Snektron** [ziglang/zig#8197]
I never did create a formal proposal issue for #5912, so here it is. Most of the important points have already been written there, but i will repeat them here for clarity.

Currently, there are implicit conversion rules for `*T` to `*[1]T`, and for `*[N]T` to `[*]T`, but there is no short-cut for `*T` to directly go  to `[*]T`. Instead, this requires writing a nasty explicit coercion using `@as` or a pointer cast.

My main motivation for this is interaction with C APIs: Because C has no run-time sized slices by default, passing an array of information usually consists of passing an unknown-length pointer and a size in a separate parameter. With C bindings that are generated by translate-c, there is no problem, as these pointers would be generated into `[*c]T`. `*T` does coerce to `[*c]T` directly. When C bindings are generated by tools with more information than translate-c (such as tools which generate bindings from specifications, like [vulkan-zig](https://github.com/Snektron/vulkan-zig), [openxr-zig](https://github.com/s-ol/openxr-zig), and [zig-wayland/](https://github.com/ifreund/zig-wayland/)), such pointers can often be generated as `[*]T` (which are more restricted than `[*c]T` and should be preferred, of course). Yet if the usage of such API often requires passing single-item pointers, writing that extra coercion gets annoying and ugly.

As a concrete example, see the amount of these coercions required in [the triangle example from vulkan-zig](https://github.com/Snektron/vulkan-zig/blob/master/examples/triangle.zig) I got lazy and used `@ptrCast`, though i have also made [coercion utilities](https://github.com/zig-community/Zig-Showdown/blob/bddb0332b1a63828ed709734a737ca8254f279ed/src/client/renderer/implementation/vulkan/util.zig#L19) in the past.

Earlier conversations with Andrew on this topic lead to the argument that while `*T` coerces to `*[1]T`, `*[2]T` does not.

> `*T` coerces to `*[1]T` which is an argument for allowing `*T` to `[*]T`, however
> `*T` does not coerce to `*[2]T` which is an argument against it
> [...]
> because `[*]T` is "unknown-length", and `*T` is length of exactly one. We wouldn't want the coercion if the lengths didn't match
> -- andrewrk

One counter-proposal also suggested by Andrew was an alternative syntax that would perform such a coercion, for example `x.many` to obtain a many-pointer from a single-item-pointer. I would also be fine with something like that.

cc @s-ol

**Comment by s-ol** [comment]
To motivate this again, here is the kind of code this applies to:
```
        const fence: vk.Fence = getFence();
        try self.vkd.resetFences(self.device, 1, @ptrCast([*]const vk.Fence, &fence));
```

here the `1` specifies the number of elements in the following unknown-size pointer.
This API style already works great with slices:
```zig
        const fences: []vk.Fence = getFences();
        try self.vkd.resetFences(self.device, fences.len, fences.ptr);
```

One counter-argument would be that two different functions/wrappers could be created here, one for slices (`resetFences()`) and one for individual pointers (`resetFence`) respectively. It is debatable whether it is a good idea to change function names such as this, mainly because this means the upstream docs are no longer correlated cleanly and there might be naming conflicts, but there are also two other cases in which there would be technical issues:

1. When the arguments are passed in a struct (with externally defined layout) this would require creating a new struct type and then copying the elements over before passing on to the actual API, which is unreasonable. Such structs are a common occurence in these APIs:
   ```zig
        const submits = [_]vk.SubmitInfo{
            .{
                .wait_semaphore_count = 0,
                .p_wait_semaphores = undefined,
                .p_wait_dst_stage_mask = undefined,
                
                .command_buffer_count = 1,
                .p_command_buffers = @ptrCast(...) // here

                .signal_semaphore_count = 0,
                .p_signal_semaphores = undefined,
            },
        };
   ```
2. When there are multiple arguments of independently-varing aritys, and so every combination of arity would have to be generated, making the API hard to understand and use:
  ```zig
        // 2^3 = 8 possible combinations
        // what would you even call them?
        self.vkd.cmdPipelineBarrier(
            buf,
            .{ .transfer_bit = true }, .{ .transfer_bit = true }, .{},
            1, @ptrCast([*]vk.MemoryBarrier, &memory_barrier),
            1, @ptrCast([*]vk.BufferMemoryBarrier, &buffer_barrier),
            1, @ptrCast([*]vk.ImageBarrier, &image_barrier)
        );
   ```

**Comment by s-ol** [comment]
Continuing in devils-advocate mode, if we did a better job wrapping the API in zig-idioms, then we would accept slices directly without the count arguments, and so at that point #3156 would be a decent solution.

That proposal is also pending, I suspect for the same reason that this proposal is?

**Comment by LemonBoy** [comment]
This cast loses information (the pointer being single) and cannot be implicit. On the other hand #3156 retains the length and is therefore safe to do implicitly.

**Comment by s-ol** [comment]
@LemonBoy but the same loss of information exists with this existing cast:
```
// Single-item pointers to arrays can be coerced to
// unknown length pointers.
test "*[N]T to [*]T" {
    var buf: [5]u8 = "hello".*;
    const x: [*]u8 = &buf;
    expect(x[4] == 'o');
    // x[5] would be an uncaught out of bounds pointer dereference!
}
```

In fact the only way to cleanly get around this at the moment is to declare a size-one array and then have the pointer to that coerce to `[*]T`, but that's silly.

**Comment by LemonBoy** [comment]
Good point, I see that as a many-to-many conversion while `*T` to `[*]T` would be one-to-many. This safety net should be enough to stop users from shooting themselves in the foot when using unsafe APIs not using slices.

**Comment by prime31** [comment]
Direct coercion, which makes logical sense to me at least would really help out with talking to C code. A `*` is exactly a `[*]` at least in my head. Perhaps I have seen too much c/c++. If direct coercion is not explicit enough then the idea of a `ptr.many` would also be a valuable addition and solve the clunkiness of `@ptrCast`.

**Comment by mlugg** [comment]
Solved by explicit, concise syntax to go from `*T` to `*[1]T` (`ptr[0..1]`) which coerces to `[*]T`.
